skiplang/compiler/src/compile.sk

```diff
@@ -193,191 +193,200 @@ fun ensureCompatibleLLVMVersion(): void {
+      shouldDisasm = (
+        funDef ~>
+          disasm &&
+            (conf.disasmFunctions.contains(funDef.origName) ||
+              (conf.disasmAnnotated &&
+                annotationsContain(
+                  funDef.annotations,
+                  "@disasm",
+                  funDef.getPos(),
+                )))
+      );
+
+      env = OuterIstToIR.createIR(
+        context,
+        conf,
+        shouldDisasm,
+        shouldRuntimeExport,
+        outerIst,
+        converter,
+        constsProj,
+        defsProj,
+      );
+
+      !env = createOptimizedFunctions(context, env, conf);
+      defs = runCompilerPhase("native/create_asm_graph", () -> {
+        AsmOutput.createAsmDefGraph(env, conf)
+      });
+
+      runCompilerPhase("native/merge_asm_graph", () -> {
+        AsmOutput.mergeIdenticalAsmDefs(defs, conf.debug)
+      });
+
+      runCompilerPhase("native/create_asm_symbols", () -> {
+        AsmOutput.assignFinalSymbols(defs)
+      });
+
+      conf.emit match {
+      | "llvm-ir" -> compile_llvm_ir(defs, conf)
+      | "link" | "cdylib" ->
+        // TODO: Remove exports stuff once --emit=cdylib is supported.
+        wasm_exports = outerIst
+          .getFuns(context, defsProj)
+          .filter(f ->
+            annotationsContainParam(
+              f.annotations,
+              "@wasm_export",
+              f.getPos(),
+            ).isSome() ||
               annotationsContainParam(
                 f.annotations,
-                "@wasm_export",
+                "@export",
                 f.getPos(),
-              ).isSome() ||
-                annotationsContainParam(
-                  f.annotations,
-                  "@export",
-                  f.getPos(),
-                ).isSome()
-            )
-            .map(exportName);
-          compile_binary(defs, conf, wasm_exports)
-        | "sklib" ->
-          invariant_violation("--emit=sklib should have been handled earlier")
-        | _ ->
-          invariant_violation("Unsupported value for `--emit`: " + conf.emit)
-        }
-      },
-    );
+              ).isSome()
+          )
+          .map(exportName);
+        compile_binary(defs, conf, wasm_exports)
+      | "sklib" ->
+        invariant_violation("--emit=sklib should have been handled earlier")
+      | _ -> invariant_violation("Unsupported value for `--emit`: " + conf.emit)
+      }
+    },
+  );
+
+  context.unsafeGetEagerDir(backendDirName)
+}
+
+fun compile(
+  config: Config.Config,
+  context: mutable SKStore.Context,
+  fileNames: Array<String>,
+): void {
+  if (fileNames.isEmpty() && config.dependencies.isEmpty()) {
+    print_error("fatal error: no input files\n");
+    skipExit(4); // Exit code used by g++ for no input files
   };
 
+  ensureCompatibleLLVMVersion();
+
+  // FIXME: Backend initialization should _not_ peak into the config
+  // as it is likely to change across invocations.
```

Comment:
s/peak/peek

AFAIU the only use made of the config is to init OuterIstToIR, and that just uses ptrBitSize. It seems that ptrBitSize should not change between invocations, and so a reasonable plan would be to pass only that value down to OuterIstToIR. WDYT?

Reply:
Good catch!

Regarding `ptrBitSize`, it can change if you successively build for `amd64` then `wasm32`. Currently we use a separate `state.db` for each arch to prevent inconsistencies.
Regardless, I agree that passing only `ptrBitSize` instead of the whole `config` is a first step in limiting chaos.